### PR TITLE
BUG: Fix wrong length of DataFrame

### LINKED ISSUE #216: [BUG] `len(df)` returns incorrect result
### Describe the bug
`len(df)` returns incorrect result.

### To Reproduce
```Python
In [1]: import xorbits.pandas as pd

In [2]: df = pd.read_csv('user.csv')

In [3]: len(df)
Out[3]: 0

```

### PR DESCRIPTION
**Pull request by aresnow1**
## What do these changes do?

For those data with unknown shape, we return 0 when call `len(df)`, which is confused for users. In this PR, trigger execution if its shape is unknown and return the correct length. 

## Related issue number

Fixes #216 

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass

### CODE REVIEW

File: python/xorbits/core/data.py [hunk 1]
```diff
@@ -212,6 +212,23 @@ def gen():
...
+                return int(self.data._mars_entity.shape[0])
+            except (IndexError, ValueError):
+                # shape is unknown, execute it
+                run(self)
+                try:
+                    return int(self.data._mars_entity.shape[0])
+                except (IndexError, ValueError):  # happens when dimension is 0
+                    return 0
+        else:
+            raise TypeError(f"object of type '{self.data.data_type}' has no len()")
```
Comment (UranusSeven) [inline python/xorbits/core/data.py:230]: The magic method wrapper raises attribute errors when the underlying object does not have a specify magic method:
```python
raise AttributeError(
                f"'{self.data.data_type.name}' object has no attribute '{method_name}'"
            )
```

Would it be better to also raise an attribute error here?

File: python/xorbits/core/data.py [hunk 2]
```diff
@@ -212,6 +212,23 @@ def gen():
...
+        if isinstance(self.data._mars_entity, HasShapeTileable):
+            try:
+                return int(self.data._mars_entity.shape[0])
+            except (IndexError, ValueError):
+                # shape is unknown, execute it
+                run(self)
+                try:
+                    return int(self.data._mars_entity.shape[0])
+                except (IndexError, ValueError):  # happens when dimension is 0
+                    return 0
```
Comment (UranusSeven) [inline python/xorbits/core/data.py:228]: The length of a zero-dimension object is quite confusing. In numpy, an error will be raised when you call len() on a zero-dimention array:

```python
In [5]: a = np.array(1)

In [6]: a.shape
Out[6]: ()

In [7]: len(a)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[7], line 1
----> 1 len(a)

TypeError: len() of unsized object
```

Would it be better if we just throw the IndexError or ValueError here?

File: python/xorbits/core/data.py (hunk 1 above)
Comment (aresnow1) [reply python/xorbits/core/data.py:230]: This is Python's behavior:
``` Python
In [1]: class A: pass

In [2]: len(A())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 len(A())

TypeError: object of type 'A' has no len()
```

File: python/xorbits/core/data.py (hunk 2 above)
Comment (aresnow1) [reply python/xorbits/core/data.py:228]: We can catch the IndexError and raise TypeError.

**Review by UranusSeven — APPROVED** [review]
LGTM

**Review by qianduoduo0904 — APPROVED** [review]
LGTM

**Pull request merged**